core/adminserver/admin_server.go

```diff
@@ -325,16 +325,42 @@ func (s *Server) initialize(ctx context.Context, logTree, mapTree *tpb.Tree) err
 
 // DeleteDomain marks a domain as deleted, but does not immediately delete it.
 func (s *Server) DeleteDomain(ctx context.Context, in *pb.DeleteDomainRequest) (*google_protobuf.Empty, error) {
+	d, err := s.GetDomain(ctx, &pb.GetDomainRequest{DomainId: in.GetDomainId()})
+	if err != nil {
+		return nil, err
+	}
+
 	if err := s.domains.SetDelete(ctx, in.GetDomainId(), true); err != nil {
 		return nil, err
 	}
+
+	_, delLogErr := s.logAdmin.DeleteTree(ctx, &tpb.DeleteTreeRequest{TreeId: d.Log.TreeId})
+	_, delMapErr := s.mapAdmin.DeleteTree(ctx, &tpb.DeleteTreeRequest{TreeId: d.Map.TreeId})
+	if delLogErr != nil || delMapErr != nil {
+		return nil, status.Errorf(codes.Internal, "adminserver: delete log %v: %v, delete map %v: %v",
+			err, d.Log.TreeId, delLogErr, d.Map.TreeId, delMapErr)
+	}
+
 	return &google_protobuf.Empty{}, nil
 }
 
 // UndeleteDomain reactivates a deleted domain - provided that UndeleteDomain is called sufficiently soon after DeleteDomain.
 func (s *Server) UndeleteDomain(ctx context.Context, in *pb.UndeleteDomainRequest) (*google_protobuf.Empty, error) {
+	d, err := s.GetDomain(ctx, &pb.GetDomainRequest{DomainId: in.GetDomainId(), ShowDeleted: true})
+	if err != nil {
+		return nil, err
+	}
+
 	if err := s.domains.SetDelete(ctx, in.GetDomainId(), false); err != nil {
 		return nil, err
 	}
+
+	_, delLogErr := s.logAdmin.UndeleteTree(ctx, &tpb.UndeleteTreeRequest{TreeId: d.Log.TreeId})
+	_, delMapErr := s.mapAdmin.UndeleteTree(ctx, &tpb.UndeleteTreeRequest{TreeId: d.Map.TreeId})
+	if delLogErr != nil || delMapErr != nil {
```

Comment:
One of these undeletions could fail, resulting in a weird state. Also SetDelete false is called before the trees are undeleted so they could both fail to undelete. Does that also create a bad state?

Reply:
If they failed, it would create a bad state. Should I try to undelete at that point?